### ISSUE TITLE
Instantiate campaign elements (story mode)

### ISSUE THREAD

**Issue by c30zD** [eloipuertas/ES2015A#133]
**Description:**
- Load the different game entities and assign them to their respective player (user or AI)

**Expected Outcome / Acceptance Criteria:**
- Game entities are taken from a "campaign" object or map
- Camera is focused over the player's stronghold

**Estimated Time:**
2h.

**Comment by c30zD** [comment]
La part de la càmera està feta. Commit 181b515443a8f50461325cb52f8aa6ed8846a3b5

**Comment by c30zD** [comment]
**Bug:** AI Controller is not properly set during instantiation

**Description:**
The AIController class does not work properly upon game initialization. As an example, its hero character doesn't move. More generally, the AI modules do not work.

**Expected Outcome / Acceptance Criteria:**
The AI can move its hero.

**Estimated Time:**
5h

**Comment by c30zD** [comment]
Vaig aturar aquesta issue fa una setmana perquè es modificaria la IA, per la qual cosa no sé si tindrà el mateix bug o d'altres o si es resoldria amb els canvis.

**Comment by c30zD** [comment]
@4nc3str4l @MSoler Si us plau, necessito saber com funciona el MicroManager, concretament què fa que l'heroi de la IA surti a atacar l'altre heroi: quines condicions s'ha de complir?, quines funcions s'executen?

**Comment by 4nc3str4l** [comment]
Doncs el Microcromanager conté 4 agents:

1 - Explorer Agent
2 - Retreat Agent
3 - AtackAgent
4 - AssistAgent

En cada iteració tots aquests agents executen la seva funció get confidence (per cada squad)

L'agent que aporta un valor més gran de confiança, és el que pren el control del esquadró en qüestio executant la seva funció  controlUnits.

L'Attack Agent es l'encarregat de gestionar els atacks i en la funció get cofidence fa el següent:

1 . Calcula un valor que he anomenat supremaci index que bàsicament el que fa és comparar com de millor és el seu squad respecte als enemics que te en un cert radi.
1. SI el seu valor de supremacia és major que 0 suma a la seva confiança el resultat d'aquesta formula:

confiança = min(valor de supremacia, 5) \* 75.

3 si a més l'esquadró enemic te l'heroi entre ells li sumen 500 a aquesta confiança.

I despres ja crida al seu control units.

Voilà, no sé que més explicar-te sobre el tema.

**Comment by c30zD** [comment]
@jackure He pensat en allò de posar l'stronghold de la IA a l'extrem del mapa i me'n vaig adonar d'un petit problema. Com et vaig comentar, canviar la posició de l'stronghold és trivial (només és canviar una línia de codi), però el problema és que vaig fer un canvi en la lògica de la instanciació de les civilitzacions perquè l'Eloi ho va demanar. Si t'hi fixes, abans tenia pensat un joc tipus "skirmish" (que és bàsicament el que em demanes), però a un dels primers sprints, l'Eloi havia demanat de tenir cada civilització a un punt concret, en comptes dels jugadors (i, en realitat això és una campanya) i només moure la càmera per veure l'stronghold del jugador. Per entendre'ns, a una campanya, cada civilització té el seu stronghold a un punt concret del mapa i l'usuari tria quin bàndol li agrada més, en canvi, a un joc tipus "skirmish", tant és que triïs una civilització com l'altra, sempre començaràs al mateix lloc, però canvia la raça (que és el que havia fet abans i el que entenc que m'estàs demanant per la IA).

Així, doncs, no tinc cap problema a fer això perquè era la meva idea inicial i en principi no és gaire complicat, però hauré de canviar més d'una línia de codi. Tot i així, el Product Owner @eloipuertas ha de donar el vist i plau.

**Comment by gpascualg** [comment]
@c30zD Crec que el que es referia @jackure i en general el grup D, és que ara mateix els elfs i els humans surten molt junts. Parlem en termes de AI/Player (perquè ara mateix no hi ha més opcions), però realment l'ho únic que volem arribar a dir és que s'han de separar més entre ells, independentment de qui és la AI i qui el Player.

Espero ser d'ajuda =P

**Comment by jackure** [comment]
@C30zD Els jocs naturals d'inici són els dos extrems superior e inferior del mapa. Deixo al vostre criteri fer que la civilització seleccionada comerci a la part inferior (i la IA a la superior),  o definir a humans a la part inferior i elfs a la superior. 
Sempre que us sembli bé :)

**Comment by c30zD** [comment]
En el commit ed2b774d2b782e5a9e43061d9e6dd44ac42f7425 he mogut el stronghold dels elfs, però he trobat alguns problemes amb els commits anteriors. Bàsicament, al FOWManager, a la línia 224, surt una IndexOutOfRangeException. Sembla que només passa quan alguna unitat s'apropa al final del mapa.

**Comment by gpascualg** [comment]
@c30zD Pel que se la matriu de posicions del `FOWManager` és del tamany del mapa. Solament hauria de sortir excepció si una entitat queda fora del mapa

**Comment by c30zD** [comment]
@gpascualg No n'estic segur, però sí vaig provar de posar els elfs a prop del cantó superior del mini-map i també de jugar com a elf. Vaig moure un civil a prop de la part superior de l'escena i va sortir l'excepció. Evidentment, podria ser que els humans hagin estat els responsables, però quan vaig jugar com a humà, l'excepció sortia molt més sense que jo hagués mogut cap unitat. Per això crec que té a veure amb el FOWManager. Malauradament, no vaig fer cap commit d'aquella posició, però si tinc temps, ho tornaré a provar. Tot i així, és possible que tingui alguna cosa a veure amb el que dius, ja que crec que en algun moment vaig clicar a fora del mapa i també algun cop passava que la IA cridava a Unit.moveTo, però sortia una NullReference.

**Comment by gpascualg** [comment]
@c30zD Sí, el `NullRefence` de la AI el tenim present (no se si ja està arreglat), tot i que teòricament no passa de ser una excepció i no arriba ni a moure's. Ara en aquest sprint miraré el tema moviment de la AI.

Respecte l'error, es raro que salti anant solament al cantó, a menys que per algun motiu Detour l'estigui enviant una mica més enllà, però realment no té sentit que ho faci. M'ho miro, però per si de cas cito a @MSoler, que ha fet la FOW i potser sap que podria estar passant.

**Comment by c30zD** [comment]
Al commit referenciat més a dalt hi és l'error. Moure una unitat a prop del cantó superior fa que salti una excepció al FOWManager. En cas de jugar com a elf, l'error és a la línia 206.

**Comment by c30zD** [comment]
**Description:**
Elves and humans are placed very close to each other, making gameplay too short.

**Acceptance Criteria:**
Both strongholds must be placed reasonably far away from each other.

**Estimated Time Effort:**
1h

**Comment by c30zD** [comment]
@jackure Hi ha cap possibilitat de fer una miqueta més gran el cub del cantó a prop de l'stronghold humà? En aquell cantó hi és el tree1_nature_human_mini_forest. Ho dic perquè és el lloc perfecte per posar-hi un stronghold: igual que a l'Edat Mitjana, un punt alt on els humans han construït una fortificació per vigilar i defensar-se dels invasors. És un punt estratègic XD. Ara bé, al commit e3fcef3033b8faac5f428092a166988f5cdd210a vaig fer algunes proves i quant a usabilitat, hi ha algunes complicacions, ja que pot passar que l'usuari construeixi dos edificis contigus i llavors les unitats no podran passar d'un costat a l'altre. Igualment, he notat que si l'usuari juga com a elf (és a dir, que el seu stronghold no és al cantó) i mou algunes unitats cap allà, pot bloquejar el pas dels civils humans, la qual cosa té gràcia com a estratègia, però se suposa que no ha de ser així.

Una altra cosa, que de fet és estètica, per tant, no hi facis gaire cas si no vols. En un principi, havia pensat de posar l'heroi a baix, al camí que porta cap a l'stronghold. Si ho faig, la idea és que l'stronghold i l'heroi es vegin des del principi, que no hagi un espai fosc entre ells i que es pugui veure el camí, però a causa de la normalització de les coordenades dels edificis, si poso l'stronghold més a prop del camí, llavors les unitats no podran passar i, en canvi, si poso l'stronghold una miqueta més lluny, quedaria massa separat de l'heroi. Per això havia pensat de posar l'heroi just al costat de l'stronghold, però així, no es veu el camí. Realment, és una qüestió que no té gaire importància, però quedaria bé.

**Comment by c30zD** [comment]
![mapa](https://cloud.githubusercontent.com/assets/9555492/11384464/54b02898-9310-11e5-8374-2338aa677ba6.jpg)

@4nc3str4l  La civilització a prop del llac són els humans.

**Comment by c30zD** [comment]
**Description:**
Now that there are two game modes, it is necessary to open each scene and load the elements.

**Expected Outcome:**
- For every scene, load the elements and assign them to the user and AI.
- Load each scene when the previous level is completed.

**Estimated Time:**
6h

**Comment by gpascualg** [comment]
Suposo que ja ho teniu controlat amb @4nc3str4l, que si no vaig errat ell feia la carrega dels edificis directament des de l'arquitecta.

**Comment by gpascualg** [comment]
Per evitar solapaments i repetir feina vull dir

**Comment by c30zD** [comment]
No, vaig enviar-li un correu i no sé si vaig comentar alguna cosa per GitHub, però de moment no n'hem parlat.

**Comment by c30zD** [comment]
Però sí que he vist que hi són les etiquetes.

**Comment by c30zD** [comment]
Quina diferència hi ha entre les etiqeutes? És a dir, què representen les etiquetes que no tenen cap referència ni a la IA ni a l'humà, @4nc3str4l, com ara "Campaign_building_military"?

**Comment by gpascualg** [comment]
@c30zD @4nc3str4l Parleu-ne doncs perque em sembla que esteu fent el mateix de formes diferents. Ara contacto amb en cristian i l'aviso ;)

**Comment by c30zD** [comment]
En aquests moments, queda per comprovar si funciona el joc quan l'usuari perd. Ara, bé, per la resta, funciona correctament.
